### DISCUSSION TITLE
Font size changes don't resize window, interacts badly with `window-step-resize`

### DISCUSSION THREAD

**Discussion by lilyball** [ghostty-org/ghostty#2661] [Feature Requests, Ideas]
When I use the Increase/Decrease/Reset Font Size actions the font size changes but the window doesn't resize. This is very surprising. Both Terminal.app and iTerm resize the window, and my natural expectation is the window should resize to preserve the cell grid because font size changes shouldn't cause layout changes.

I filed [an issue on this](https://github.com/ghostty-org/ghostty/issues/2659) and was told this behavior is intentional, and that surprised me. Why should the font size change layout? This isn't a text editor with soft wrapping, it's a terminal where exact layout can matter a lot. If I've set the window to a specific row/column count, it should stay at that count until I tell it otherwise. Many programs don't handle resizing well while running (e.g. if I'm reading a `man` page it lays out text to match my terminal width, but it doesn't relayout on resize, so if I want to increase font size for readability I need to quit and re-run the `man` command). Or maybe I'm in a shared `zellij` session and increasing font size suddenly caused my code pairing partner's zellij pane to shrink.

Beyond the arguments for why the behavior should be changed, the current behavior interacts badly with `window-step-resize`. The point of `window-step-resize` is to keep the window aligned with the cell grid so there's no extra padding. But when I change font size and the window doesn't resize, the window loses its alignment with the cell grid and this causes additional padding. And because `window-step-resize` works by just forcing resizing to work in increments of the cell size rather than snapping to the cell grid, there's no way to resize my window to get rid of that padding.

Here's some steps to try to see this:

1. Ensure `window-step-resize` is on
2. Increase font size a bunch
3. Resize the window slightly
4. Reset font size
5. Try to resize the window back to its original size

Unless you're very lucky, the window resizing is now operating on a different grid than before and you cannot match the original pixel dimensions of the window, which also means you're stuck with the extra padding, unless you repeat these steps in reverse.

The `window-step-resize` issue could be worked around by either snapping the window back to the cell grid the moment the user tries to resize, or snapping it to the cell grid when changing font size (but having the window pixel dimensions waver slightly as the font size changes would be weird), but really this shouldn't happen at all because font size changes should simply resize the window to preserve the cell layout.

**Comment by lilyball** [comment]
Oh another quirk of the current behavior: if I set `window-padding-balance = true`, then changing the font size also changes the left/top padding so all of the text jumps around oddly.

**Comment by mitchellh** [comment]
> I filed https://github.com/ghostty-org/ghostty/issues/2659 and was told this behavior is intentional, and that surprised me. Why should the font size change layout? This isn't a text editor with soft wrapping, it's a terminal where exact layout can matter a lot.

The decision to make it intentional was mine. My previous terminal was Kitty and it behaved this way and I liked this behavior. Other terminals also behave this way (Alacritty another). I don't think there's any consensus on correctness here, I think it's a matter of personal taste. In my opinion, the window changing size is worse, more software handles layout just fine on resize than not, etc. But, I can see the reverse argument too.

**I will not be changing this behavior, but I am open to introducing a configuration or keybind that behaves the way you request.** Because, as I said, I can see the argument for it, and I'm not sure I've ever seen a terminal make this configurable before. A keybinding would be particularly interesting because then you can have _both_ at the same time (for situations you're using a program that doesn't relayout well).

I think there is also merit that this behavior should change in the context of certain other configurations such as `window-step-resize`. Namely, if that is true, we should probably default to font changes also changing the window size. I don't think that'd be surprising and I'm supportive of that. But supporting window size changes on font size change is probably the first thing that should be done.

To move that forward, if there is interest, I think some thought needs to go into how this would work with our apprt system. Presently, font change is purely libghostty and apprts don't even know it happens.

**Reply by poige** [reply]
> and I'm not sure I've ever seen a terminal make this configurable before

![image](https://github.com/user-attachments/assets/bf589b94-1ff6-449c-adf7-cf947925e555)

https://apple.stackexchange.com/a/222329

**Comment by lilyball** [comment]
I would be ok with the idea that `window-step-resize` controls the behavior here, since `window-step-resize` is basically the "I want my window size to sync to the cell grid" option, and what I'm asking for here is for the window size to simply reflect the correct size for my cell grid.

Also to note, if I use a configuration change that switches fonts, that too is a spot where the window size needs to adjust itself for the new cell grid when using `window-step-resize`.

**Reply by mitchellh** [reply]
It really needs to happen I think for any font property change that could affect the grid. We luckily have a hashed value that can be used to detect this already (`SharedFontGridSet.Key`) since that is how we dedupe memory usage across terminals that share identical font settings (this was a source of extreme memory bloat once upon a time). 

To be explicit: there are many more things that affect the grid that are runtime-changeable such as `adjust-cell-width` and so on. And there could be even more in the future. These are all tied to the font grid key I _think_ because they affect things like sprite rendering (implemented in the font system). Confirmation needed.

So I think what we need to do is detect anytime the font grid key changes, recalculate a desired screen size, and ask the apprt to go to that size. If the apprt _fails_ or _doesn't support_ going to that size, we can fallback to resizing the grid.

**Comment by seanh** [comment]
I came here to report the same issue, and found this discussion.

:+1: to having the behaviour change when `window-step-resize` is on. `window-step-resize` currently breaks when you change the font size and I don't think there's any solution to that other than changing the window size when the font size changes, at least when `window-step-resize` is on.

* * *

I use `window-step-resize` because I don't want a random amount of padding at the edges of the window because I think that looks bad, I want things drawn in the terminal with a background color (like status bars etc) to go right to the edge of the window. Currently `window-step-resize` achieves this beautifully but changing the font size breaks it.

Also second @lilyball's point that keeping the number of rows and columns constant can be desirable to avoid messing up what's currently displayed in the terminal.

I do agree that this may be one of those personal taste things about terminals, where different users just have different preferences and there's no one behavior that can satisfy everyone. So I think tying this to `window-step-resize` makes sense.

**Comment by JakeWharton** [comment]
Related https://github.com/ghostty-org/ghostty/discussions/2946

**Comment by ReArmedHalo** [comment]
I would also like to see this feature. I'm a legally blind user, just getting started with Ghostty (loving it so far) and the first thing I do on any terminal is increase the font size to something more easily readable for myself without needing the system magnifier enabled (i am a macOS user).

I set my terminal size usually between 18 and 24 pts so with the default window size things can get pretty cramped. iTerm has this window resizing mechanic by default. On occasion, I will need to temporarily increase my font to read some text better, or during a screen sharing session, and having to resize the window gets messy with certain prompts or command outputs (I have to deal a lot with PowerShell table outputs from Microsoft 365 services).

I'm overcoming this for now by setting the `window-height` and `window-width` config options. I saw options to start maximized or fullscreen but that isn't always desired either.

+1 to having a configuration option be added to adjust the window with the font.

**Comment by johnsonjh** [comment]
Yes, I'm trying out various terminals and ghostty would fit the bill for me, but the window resize behavior is a deal breaker.

I use many legacy applications that require a fixed terminal size (80x24, 132x78, etc.) and resizing them corrupts all the text and requires restarting, and it also makes man and many other common commands completely useless.  I'd love to have an option to disable this behavior!

**Comment by majick** [comment]
The fact that this is intentional is stunningly baffling and I do hope a contributor is able to overcome this apparently fundamental design decision. Effectively all of the other opinionated default behaviors that make ghostty feel like a term-ricing project from the 90s can be fixed trivially: `macos-titlebar-style`, `window-theme`, that sort of thing, and it's even fairly easy to undo the heavy emphasis on tabs (which is obviously more a matter of individual workflow than taste.) But for some reason this, of all possible opinionated behaviors, seems to be the hill to die on.

Forcing the tty's dimensions to change every time the font changes is an odd choice for an application that seems to be so thoughtful about tty dimensions everywhere else; no other action that I could see alters the dimensions of the tty itself without an intentional choice on the part of the user. Apart from the nifty split feature, that is, but those are a bit uncommon to need at the terminal emulation layer now that it's not 1986 and `screen` has existed for almost 40 years.

I'll be keeping an eye on the project, for sure, in hopes this becomes fixable. Spending an hour-ish on getting it set up almost felt satisfyingly worthwhile until I tried to zoom a window for someone to see over my shoulder, at which point I found an egregious "bug." It's a damned nice piece of software and a really cool project. I've been really looking forward to making it a primary tool. A single misfeature was all it took, though, and that's a real shame.

My hope is that someone with as strong an opinion of this as I have will be able to make a contribution here. Until then, thanks to everyone who has put so much work into it! The day I can actually deploy the software will be a banner day for me — nearly everything else out there doesn't come close.

**Reply by poige** [reply]
It's in Zig. Do you speak it? ©

**Comment by seanh** [comment]
> **I will not be changing this behavior, but I am open to introducing a configuration or keybind that behaves the way you request.** (@mitchellh)

I think this suggestion blew by me when I first read it but I want to give a :+1: to Mitchell's suggestion here to **add a new pair of keybindings for the requested behavior**.

So the existing `increase_font_size` and `decrease_font_size` actions would retain their current behavior and would continue to be bound to `super+equal` and `super++` by default. Then add new `increase_font_and_window_size` and `decrease_font_and_window_size` actions (those probably aren't the ideal names for them) that need not have any default keybinds. Those of us who prefer that behavior can then just change the keybinds in our configs.

This approach would also allow a user to have both: you could have keybinds for _both_ pairs of actions (or use the command palette). While I generally want the window dimensions in columns and rows to remain fixed when changing my font size, I can also see the argument for the current behaviour: sometimes I've arranged my windows on screen just-so and don't want to mess up that layout, and sometimes I _don't_ actually care if the number of columns and rows changes. So I could imagine myself using both pairs of actions at different times.
